### DISCUSSION TITLE
Where is the cx_freeze script?

### DISCUSSION THREAD

**Discussion by ThomasFreedman** [marcelotduarte/cx_Freeze#2039] [Q&A] [answered]
Just starting out trying to create a binary executable for my python GUI app on Linux using cx_freeze. The documentation is a bit vague, saying the cx_freeze script is *is included with other Python scripts*, and on Unix it is in *the bin directory of the prefix where Python is installed*.

Not sure what is meant by *the prefix where Python is installed*, but I used find + grep to locate it and came up with nothing. 

Also note I am installing into a virtualenv and it wasn't under that folder hierarchy either.

There is so much flooding of info about cx_freeze for Windoze that it's not easy to find it for other platforms.

Looking for a quickstart guide, youtube video or info on how to use this tool.  I presume the cx_freeze script is a helper script to simplify use, but without that I'm looking at the `https://cx-freeze.readthedocs.io/en/stable/setup_script.html#setup-script`.

Reading there I see info about "installers" for mac & win, but only rpm for Linux. In general it seems less info is provided for the Linux platform. My target OS is a debian derrivative (MX-Linux). 

I have read where cx_freeze doesn't produce a 1 file executable, but since that is the case I'd like to see just how an app is to be used on Linux. If a .deb package is created that would be good, but not seeing a description about the options for installation on Linux.

I attempted to use the setup.py approach which produced an executable and "zillions" of things in the "lib" folder. I ran the executable but it failed, most likely due to the dynamically imported items in one of my 3rd party modules. Since my app is using tkinter, I removed the exclude line in setup.py `["tkinter", "unittest"]`. Since it couldn't find tkinter I'm guessing that line wasn't only for "tkinter unit tests".

Assuming I can identify all of the dynamically imported modules, where do I list them in setup.py?

**Comment by marcelotduarte** [comment, marked as answer]
If you read [Using cx_Freeze](https://cx-freeze.readthedocs.io/en/stable/overview.html), you can use cxfreeze or setup script.
The name of script is [cxfreeze](https://cx-freeze.readthedocs.io/en/stable/script.html).
The script is cxfreeze not cx_freeze. The project's name is cx_Freeze but the name of the command line command is cxfreeze.

> I have read where cx_freeze doesn't produce a 1 file executable, but since that is the case I'd like to see just how an app is to be used on Linux. If a .deb package is created that would be good, but not seeing a description about the options for installation on Linux.

cx_Freeze can produce one or more executables.

We cannot create .deb, but I hope that you can convert the rpm to deb or rpm to tgz using alien.

`sudo apt-get install alien`

In the future, we can have [appimage](https://appimage.org/) or other.

> I attempted to use the setup.py approach which produced an executable and "zillions" of things in the "lib" folder. I ran the executable but it failed, most likely due to the dynamically imported items in one of my 3rd party modules. Since my app is using tkinter, I removed the exclude line in setup.py ["tkinter", "unittest"]. Since it couldn't find tkinter I'm guessing that line wasn't only for "tkinter unit tests".

If you use [Setup script](https://cx-freeze.readthedocs.io/en/stable/setup_script.html), you can change according.
The are various sample [here](https://github.com/marcelotduarte/cx_Freeze/tree/main/samples).

> Assuming I can identify all of the dynamically imported modules, where do I list them in setup.py?

Use [includes](https://cx-freeze.readthedocs.io/en/stable/setup_script.html#cmdoption-arg-includes) or [packages](https://cx-freeze.readthedocs.io/en/stable/setup_script.html#cmdoption-arg-packages) options.

**Reply by ThomasFreedman** [reply]
Thank you! Doh! on using wrong name (cx_freeze), I see cxfreeze now.

I have been playing around using the setup.py generated with `cxfreeze-quickstart` and making some progress. I seem to recall running into a similar issue when I tried to use pyInstaller, that being the way one of the 3rd party libraries does relative imports. Here is the snippet from the 3rd party lib:

```
if get_platform() == "android":
    from .Interfaces import Interface
    from .Interfaces import LocalInterface
    from .Interfaces import AutoInterface
    from .Interfaces import TCPInterface
    from .Interfaces import UDPInterface
    from .Interfaces import I2PInterface
    from .Interfaces.Android import RNodeInterface
    from .Interfaces.Android import SerialInterface
    from .Interfaces.Android import KISSInterface
else:
    from .Interfaces import *
```
and the line producing the error:
`    interface_mode = Interface.Interface.MODE_FULL`
`NameError: name 'Interface' is not defined`

Not sure what to do to resolve. I've made several attempts to add "Interfaces" in various ways to setup.py's *includes* , with and without the parrent package (which is RNS). Here's the last thing I tried:

includes = "RNS.Interfaces,RNS.Interfaces.Interface"

I also had trouble with case sensitivity for the NotifyPy library. The original source was:
`from NotifyPy import Notify`
cxfreeze said "notify" could not be found. The name of the library/package (what is the diff in this cxfreeze context?) in site-packages is "notifypy, which differs from original source which works. I had to change the source to:
`from notifypy import notify as note`
and the previous reference to Notify to note, which resolved the issue (at least the error disappeared).

Hoping to get past this RNS Interfaces problem, and discover that's the last one.

Thanks for the suggestion to convert rpm to .deb, I'll definitely look into that. As for a tgz, what would the recipient of that need to do to run the program? Not so good for a non-dev type of user, but if the untared file resembles the exe folder at least a script could easily be written to untar the file and run the executable in it.

**Comment by marcelotduarte** [comment]
>  and the line producing the error:
>     interface_mode = Interface.Interface.MODE_FULL
> NameError: name 'Interface' is not defined

What is the name of the package?

> I also had trouble with case sensitivity for the NotifyPy library. The original source was:
> from NotifyPy import Notify

Are you using https://pypi.org/project/notify-py/ ?

The simple usage is:
`from notifypy import Notify`

So the package name is notify-py but the import name is notifypy (which is used by Python and cx_Freeze).

Tip: To build a app as an executable, you should test if it works in Python first. If it runs without errors, you can convert it to executable.

**Reply by ThomasFreedman** [reply]
Thx again.

No, the notify package is NotifyPy, which is different than notify-py. The original source import spelling and case work, if I just run the program in the same venv folder I'm running setup.py in. It also has no issues with the RNS library/package, the GUI app starts just fine. 

The name of the package producing the Interfaces error is RNS.

**Comment by marcelotduarte** [comment]
> The name of the package producing the Interfaces error is RNS.

packages = ["RNS"] should work.

**Reply by ThomasFreedman** [reply]
I'll give that a try. Nope, produces the exact same error.

**Reply by ThomasFreedman** [reply]
```
  File "pboxSearch.py", line 51, in initialize
    pbq = qb.pBoxQuery(config)          # Creates all object classes
  File "/volumes/Accounts/owner/PycharmProjects/pBoxV2/MediaCaptureForm/test/queryBuilder.py", line 24, in __init__
    self.Ret      = rt.RetComm(self)
  File "/volumes/Accounts/owner/PycharmProjects/pBoxV2/MediaCaptureForm/test/reticulum.py", line 141, in __init__
    RNS.Reticulum(self.RetCfgFile)    # Initialize the Reticulum Network Stack
  File "/volumes/Accounts/owner/PycharmProjects/pBoxV2/MediaCaptureForm/test/lib/python3.9/site-packages/RNS/Reticulum.py", line 255, in __init__
    self.__apply_config()
  File "/volumes/Accounts/owner/PycharmProjects/pBoxV2/MediaCaptureForm/test/lib/python3.9/site-packages/RNS/Reticulum.py", line 391, in __apply_config
    interface_mode = Interface.Interface.MODE_FULL
NameError: name 'Interface' is not defined

```

**Reply by ThomasFreedman** [reply]
As suspected the problem is with this 3rd party package, RNS.

It may be useful for you to discover why cx_freeze fails to process this package's imports without a run time error. I have created a very simple console program that attempts to initialize the RNS package that exhibits the problem:

```
#!/usr/bin/python3
import RNS
RNS.Reticulum()           # Initialize the Reticulum Network Stack
print("Reticulum object construction is complete")
```

Here is the setup.py I used to build the program:

```
from cx_Freeze import setup, Executable

# Dependencies are automatically detected, but it might need
# fine tuning.
build_options = {'packages': ['RNS'], 'excludes': []}

base = 'console'

executables = [
    Executable('tstRns.py', base=base)
]

setup(name='tstRns',
      version = '0.1',
      description = 'Test to only init RNS',
      options = {'build_exe': build_options},
      executables = executables)
```

The context is a new venv using Python 3.9.2 where only the RNS package is installed with pip3.

UPDATE:
The issue is dynamic imports as suspected. I modified the RNS.Reticulum.py file to assume a non-Android platform which eliminated the run time error and produced a functional program.

I'm not sure what the limits are for detecting & accomodating dynamic imports are in cx_freeze or if it's worth your time to address them. It seems to me there is a huge range of possibilities that would be extremely difficult to detect. 

I am now going to see if cx_freeze will work for a more complicated GUI app using the modified RNS package. If so I can proceed, and will raise the issue with the RNS author.

Thanks so much for your help & suggestions on using cx_freeze.

**Reply by marcelotduarte** [reply]
Please open an [issue](https://github.com/marcelotduarte/cx_Freeze/issues/new?assignees=&labels=&projects=&template=bug-report.md&title=) with this information. I'll check and can create hooks to make it work.

**Reply by ThomasFreedman** [reply]
Thank you sir, I just submitted issue #2041 as suggested.

**Comment by ThomasFreedman** [comment]
I have used alien as you suggested to convert the rpm package created by cx_freeze to a debian .deb file, which installs without error using dpkg -i, however all the files are those in the venv folder, none are installed to /bin, /usr/bin or any other standard location in my PATH. 

I also installed the package on a newly installed MX-Linux system, and discovered it didn't actually do a functional install as I would expect a .deb package to do, rather it recreated my entire folder hierarchy with my account name. 

I don't know if that is how the rpm variant works, but if so and for debian, cx_freeze doesn't produce an   end-user package. Is this as intended? An appImage was mentioned in another thread, which would be good. So would an rpm or deb file if it were made to install things so they can be executed irrespective of any user account.

**Comment by marcelotduarte** [comment]
I'll try to find some time to check the rpm (and alien usage) to see the issues you raised.
I already have a working bdist_appimage, it lacks tests and documentation. I can invite you to do tests in a few days.

**Reply by ThomasFreedman** [reply]
I'll be happy to help test it, thanks again. After seeing some progress with cx_freeze and discovering a way to work around the dynamic imports of the RNS package, I decided to revisit pyinstaller.

I got one of my 2 GUI apps to mostly work, but will need to change the code a bit to use pyinstaller, which is doable but not preferable. However, the 2nd app still has a problem with RNS despite my change to eliminate the dynamic import issue that caused the problem in the first app. I need to double check my work tho, as I suspect the 2nd build is pulling RNS from outside the venv.

I was reading about the nature of appImages a month ago or so, and my take away was it's usefullness is slipping away as operating systems evolve. I don't recall the details now, but my impression is that due to the appImage architecture and the way OSs are evolving, appImages will become less "universal" and more problematic in the future. 

I really do like their current "exe", run anywhere capability tho, just like Micro$oft .exe files provide for Windows. I think the "onefile" approach is the best way to go. 

Thanks again for all you do in bringing tools like cx_freeze to the public!

**Comment by marcelotduarte** [comment]
> I was reading about the nature of appImages a month ago or so, and my take away was it's usefullness is slipping away as operating systems evolve. I don't recall the details now, but my impression is that due to the appImage architecture and the way OSs are evolving, appImages will become less "universal" and more problematic in the future.

I also have some of these impressions. I think snap will be used more, then I think about implementing it too, but appimage is easier to implement. There are several implementations for distributing Python code (for example python-appimage using manylinux images). As manylinux images are used, some problems are avoided.
So I did it in the same vein, I also use python for manylinux images, so generating an appimage was easy.

> I think the "onefile" approach is the best way to go.

This has a problem, the app is slow to load and sometimes to run. I'm not talking about appimage, but onefile or putting everything in zip. The appimage still needs to be tested with a larger project.

**Reply by ThomasFreedman** [reply]
I'm not a fan of snap. I don't like its' overhead and that snap apps impose an entirely separate subsystem to deal with them. I avoid using snap images.

I'm not familiar with manylinux images. I'll look into that.

**Reply by ThomasFreedman** [reply]
Life is all about choices & tradeoffs. The concern I have with manylinux is the tradeoff between security and the convenience of a universal linux executable file.

When it comes to cyber security I believe there are many undiscovered vulnerabilities in all computer code. How those vulnerabilities surface (if they ever do) depend on too many variables to provide confidence all are covered and handled. 

Thus the bottom line for me is the difficulty of evaluating the security risks for each type of packaging & distribution.

**Comment by marcelotduarte** [comment]
> The concern I have with manylinux is the tradeoff between security and the convenience of a universal linux executable file.

Manylinux have various versions. For manylinux2014 that I use, the security updates are applied before normal distributions. Comparing to Ubuntu 22.04:
Ubuntu - > Manylinux
Python 3.10.12 -> 3.10.13  (both are security fixes, from June and August, respectively)
OpenSSL 3.0.2 -> 1.1.1w (1.1.1w has more security updates than 3.0.2, and we'll upgrade to 3.0.10 after some tests in 2 weeks)
Tkinter 8.6.12 -> 8.6.13
...

**Reply by ThomasFreedman** [reply]
Thanks for keeping me informed. I've gotten bogged down in some other things & haven't gotten back to testing with cx_freeze. I will though & I will definitely let you know if your fix was able to pick up conditional imports in the RNS package that were causing problems. I truly appreciate your work and your quick response to the issue I raised.
